packages/flame/lib/src/components/navigator_component.dart

```diff
@@ -0,0 +1,206 @@
+import 'package:flame/src/components/component.dart';
+import 'package:flame/src/components/route.dart';
+import 'package:meta/meta.dart';
+
+/// [NavigatorComponent] handles transitions between multiple pages of a game.
+///
+/// The term **page** is used descriptively here: it is any full-screen (or
+/// partial-screen) component. For example: a starting page, a settings page,
+/// the main game world page, and so on. A page can also be any individual piece
+/// of UI, such as a confirmation dialog box, or a user inventory pop-up.
+///
+/// The navigator doesn't handle the pages directly -- instead, it operates
+/// a stack of [Route]s. Each route, in turn, manages a single page component.
+/// However, routes are "lazy": they will only build their pages when they
+/// become active.
+///
+/// Internally, the Navigator maintains a stack of Routes. In the beginning,
+/// the stack will contain the [initialRoute]. New routes can be added via the
+/// [pushNamed] method, and removed with [pop]. However, the stack must be
+/// kept non-empty: it is an error to attempt to remove the only remaining route
+/// from the stack.
+///
+/// Routes that are on the stack are mounted as components. When a route is
+/// popped, it is removed from the stack and unmounted. Routes can be either
+/// transparent or opaque. An opaque route prevents all routes below it from
+/// rendering, and also stops pointer events. In addition, routes are able to
+/// stop or slow down time for the pages that they control, or to apply visual
+/// effects to those pages.
+class NavigatorComponent extends Component {
+  NavigatorComponent({
+    required this.initialRoute,
+    required Map<String, Route> routes,
```

Comment:
What was the verdict of making this `Map<T, Route>` so that enums for example can be used too?

Reply:
The advantage of string ids is that they are an open-ended set: the route names can be generated at runtime. For example, if the game includes some kind of downloadable levels, then it needs a way to invoke those levels from the Navigator. For this purpose we have `routeFactories` and `pushRoute()`. It wouldn't be possible to implement route factories with generic `T` (unless we specify that those method work only when `T = string`).

Also, it would be harder to create routes on the fly: you'd have to write `Route<MyNavigatorIds>(...)` instead of `Route(...)`.

At the same time, the benefit that the enums would bring into the Navigator is better type safety: it would be impossible to accidentally misspell the name of a route. However, the same safety could be achieved by simply declaring a number of static strings for each route and using those strings as names. This would work as a "manual enum" without sacrificing any of the other Navigator's functionality.

Reply:
The generics can't be omitted since they don't have the same base class, but setting it to `String` would work, but the problem is that it would have to be done everywhere all the time if we don't have an implementation for just strings like`class StringRoute extends Route<String>`.

It would be nice to allow other types than string, I find manual enums a bit dirty. But I also don't want to sacrifice the beginner's user experience in favor of this.

Wdyt of allowing `T` in the base class, but having ready implementations for `String` as key, so that the user doesn't have to specify it everywhere?

Reply:
> The generics can't be omitted since they don't have the same base class

Hm, I am not sure about that, I just tried this on dart pad to confirm and it worked nice, `value` is dynamic because no generics was informed:

```dart
class Value<T> {
  T? value;
}

void main() {
  final v = Value();
  v.value = 'A';
  print(v.value);
}
```

Am I missing something obvious?

Reply:
Sure, but then it wouldn't be typed at all, the worst of all worlds :smile: